toolchain/check/pattern_match.cpp

```diff
@@ -0,0 +1,193 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/check/pattern_match.h"
+
+#include <functional>
+#include <vector>
+
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SmallVector.h"
+#include "toolchain/base/kind_switch.h"
+#include "toolchain/check/context.h"
+#include "toolchain/check/convert.h"
+
+namespace Carbon::Check {
+namespace {
+
+// The collected state of a pattern-matching operation.
+class MatchContext {
+ public:
+  struct WorkItem {
+    SemIR::InstId pattern_id;
+    // Invalid when processing the callee side.
+    SemIR::InstId scrutinee_id;
+  };
+
+  // Constructs a MatchContext. If `callee_specific_id` is valid, this pattern
+  // match operation is part of implementing the signature of the given
+  // specific.
+  explicit MatchContext(
+      SemIR::SpecificId callee_specific_id = SemIR::SpecificId::Invalid)
+      : result_(SemIR::InstId::Invalid),
+        callee_specific_id_(callee_specific_id) {}
+
+  // Returns whether there are any work items to process.
+  auto HasWork() const -> bool {
+    return !stack_.empty() && !result_.is_valid();
+  }
+
+  // Adds a work item to the stack. Cannot be called after Finish().
+  auto AddWork(WorkItem work_item) -> void {
+    CARBON_CHECK(!result_.is_valid());
+    stack_.push_back(work_item);
+  }
+
+  // Returns the next work item to process.
+  auto NextWorkItem() -> WorkItem { return stack_.pop_back_val(); }
+
+  // Sets the result of this pattern matching operation. Must not be called
+  // when there is still pending work, except to report an error.
+  auto Finish(SemIR::InstId result) -> void {
+    CARBON_CHECK(!HasWork() || result == SemIR::InstId::BuiltinError);
+    result_ = result;
+  }
+
+  auto result() const -> SemIR::InstId { return result_; }
+
+  auto callee_specific_id() const -> SemIR::SpecificId {
+    return callee_specific_id_;
+  }
+
+ private:
+  llvm::SmallVector<WorkItem> stack_;
+
+  SemIR::InstId result_;
+
+  SemIR::SpecificId callee_specific_id_;
+};
+
+auto ProcessPattern(Context& context, MatchContext& match,
+                    MatchContext::WorkItem entry) {
+  auto pattern = context.insts().GetWithLocId(entry.pattern_id);
+  CARBON_KIND_SWITCH(pattern.inst) {
+    case SemIR::BindingPattern::Kind:
+    case SemIR::SymbolicBindingPattern::Kind: {
+      CARBON_CHECK(!entry.scrutinee_id.is_valid());
+      auto binding_pattern = pattern.inst.As<SemIR::AnyBindingPattern>();
+      auto bind_name = context.insts().GetAs<SemIR::AnyBindName>(
+          binding_pattern.bind_name_id);
+      context.inst_block_stack().AddInstId(bind_name.value_id);
+      context.inst_block_stack().AddInstId(binding_pattern.bind_name_id);
+      match.Finish(binding_pattern.bind_name_id);
+      break;
+    }
+    case CARBON_KIND(SemIR::AddrPattern addr_pattern): {
+      if (!entry.scrutinee_id.is_valid()) {
+        // We're still on the caller side of the pattern, so we traverse without
+        // emitting any insts.
```

Comment:
This `if` condition + comment is a bit confusing, because the `if` condition is "are we matching on the callee side" and the comment seems to say we're on the caller side. I've figured out what you mean, but to make this clearer to future readers, how would you feel about something like:
```suggestion
      if (!entry.scrutinee_id.is_valid()) {
        // We're emitting match IR for the callee but we're still on the caller side of the pattern,
        // so we traverse without emitting any insts.
```

I also think it might be clearer to add a `bool` to `MatchContext` to indicate whether we're emitting match IR for the caller or callee side, rather than tracking it on the `WorkItem` (which suggests it might change between work items). Maybe a CHECK at the start of this function that we have a scrutinee iff we're on the caller side would be good.

Reply:
OK, how's this?

> I also think it might be clearer to add a `bool` to `MatchContext` to indicate whether we're emitting match IR for the caller or callee side, rather than tracking it on the `WorkItem` (which suggests it might change between work items).

Good point, but I made it an `enum` because we eventually want to support non-function-call pattern matches as well.

> Maybe a CHECK at the start of this function that we have a scrutinee iff we're on the caller side would be good.

I don't intend that assertion to hold in the long run, so I'd rather not assert it now. For example, as noted on the previous comment, I intend for the `*BindingPattern` case to have a scrutinee eventually (hopefully the PR after this one). The invariant that I'm aiming for is that we have a scrutinee iff we should emit match insts, but we aren't quite there yet, so I've marked that as a TODO. In any event, that `CHECK` will need to be applied to the individual cases, because there's no top-level representation of "should emit match insts".